iis-wf/iis-wf-referenceextraction/src/main/java/eu/dnetlib/iis/wf/referenceextraction/patent/PatentMetadataRetrieverJob.java

```diff
@@ -52,50 +63,101 @@
+            
+            Configuration hadoopConf = sc.hadoopConfiguration();
+            
+            HdfsUtils.remove(hadoopConf, params.getOutputPath());
+            HdfsUtils.remove(hadoopConf, params.getOutputFaultPath());
+            HdfsUtils.remove(hadoopConf, params.getOutputReportPath());
+            
+            LockManager lockManager = LockManagerUtils.instantiateLockManager(params.getLockManagerFactoryClassName(),
+                    hadoopConf);
             
             try {
                 PatentServiceFacade patentServiceFacade = ServiceFacadeUtils
                         .instantiate(prepareFacadeParameters(params.patentFacadeFactoryClassname, params.facadeParams));
                 
-                JavaRDD<ImportedPatent> importedPatents = avroLoader.loadJavaRDD(sc, params.inputPath,
-                        ImportedPatent.class);
+                JavaRDD<ImportedPatent> importedPatents = avroLoader.loadJavaRDD(sc, params.inputPath, ImportedPatent.class);
+                
+                final String cacheRootDir = CacheStorageUtils.normalizePath(params.getCacheRootDir());
+                CacheMetadataManagingProcess cacheManager = new CacheMetadataManagingProcess();
+                
+                String existingCacheId = cacheManager.getExistingCacheId(hadoopConf, cacheRootDir);
+                
+                // skipping already extracted
+                JavaRDD<DocumentText> cachedSources = CacheStorageUtils.getRddOrEmpty(sc, avroLoader, cacheRootDir,
+                        existingCacheId, CacheRecordType.text, DocumentText.class);
+                // caching: will be written in new cache version and output
+                cachedSources.cache();
+                
+                JavaPairRDD<CharSequence, DocumentText> cacheById = cachedSources.mapToPair(x -> new Tuple2<>(x.getId(), x));
+                JavaPairRDD<CharSequence, ImportedPatent> inputById = importedPatents.mapToPair(x -> new Tuple2<>(getId(x), x));
+                JavaPairRDD<CharSequence, Tuple2<ImportedPatent, Optional<DocumentText>>> inputJoinedWithCache = inputById.leftOuterJoin(cacheById);
 
-                JavaPairRDD<CharSequence, ContentRetrieverResponse> obtainedIdToResponse = importedPatents
-                        .mapToPair(x -> new Tuple2<>(getId(x), getMetadataFromFacade(x, patentServiceFacade)));
-                obtainedIdToResponse.persist(StorageLevel.DISK_ONLY());
+                JavaRDD<ImportedPatent> toBeProcessed = inputJoinedWithCache.filter(x -> !x._2._2.isPresent()).values().map(x -> x._1);
+                JavaRDD<DocumentText> entitiesReturnedFromCache = inputJoinedWithCache.filter(x -> x._2._2.isPresent()).values().map(x -> x._2.get());
+                entitiesReturnedFromCache.cache();
+                
+                JavaPairRDD<CharSequence, ContentRetrieverResponse> returnedFromEPO = retriveFromRemoteEndpoint(toBeProcessed, patentServiceFacade);
+                returnedFromEPO.cache();
```

Comment:
Same response can be applied here. What is more, knowing the caches are updated incrementally the metadata returned from EPO would be much smaller than the total cache size because it will hold newly matched patents only.